### PR TITLE
Add support for Input Text and Modal components

### LINKED ISSUE #490: Add Support for Modals & Input Text in Application Commands
https://github.com/discord/discord-api-docs/pull/4143

### PR DESCRIPTION
**Pull request by krittick**
## Summary

Remake of #573 because I broke things in my fork.

**Note:** Pretty much done, works as expected. Might need some more work on docstrings.

**Checklist:**

- [x] Base objects/classes/enums
  - [x] Input Text
  - [x] Modals
- [x] Docstrings
  - [x] Input Text
  - [x] Modals
- [x] Examples to show usage

This adds support for input text and modal components.

Closes #490 

## Checklist

- [ ] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [x] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

Credit to @Chiggy-Playz for their guidance in pointing me toward re-using many of the methodologies from View/Button/Select to quickly get our implementation running.

### CODE REVIEW

**Review by Lulalaby — DISMISSED** [review]

**Comment by BobDotCom** [comment]
Placed this under 2.1 since it might not release before 2.0 comes out.

**Review by Lulalaby — APPROVED** [review]

**Review by Dorukyum — DISMISSED** [review]

**Review by Lulalaby — DISMISSED** [review]

**Comment by Lulalaby** [comment]
⚠️ Do not merge yet

File: examples/modal_dialogs.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+
+@bot.slash_command(name="modaltest", guild_ids=[907533384081879040])
+async def modal_slash(ctx):
+    """Shows an example of a modal dialog being invoked from a slash command."""
+    modal = MyModal()
+    await ctx.interaction.response.send_modal(modal)
+
+
+@bot.message_command(name="messagemodal", guild_ids=[907533384081879040])
```
Comment (Middledot) [inline examples/modal_dialogs.py:41]:
```suggestion
@bot.message_command(name="messagemodal", guild_ids=[...])
```

File: examples/modal_dialogs.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+async def modal_message(ctx, message):
+    """Shows an example of a modal dialog being invoked from a message command."""
+    modal = MyModal()
+    modal.title = f"Modal for Message ID: {message.id}"
+    await ctx.interaction.response.send_modal(modal)
+
+
+@bot.user_command(name="usermodal", guild_ids=[907533384081879040])
```
Comment (Middledot) [inline examples/modal_dialogs.py:49]:
```suggestion
@bot.user_command(name="usermodal", guild_ids=[...])
```

File: examples/modal_dialogs.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        )
+
+    async def callback(self, interaction: discord.Interaction):
+        embed = discord.Embed(title="Your Modal Results", color=discord.Color.random())
+        embed.add_field(name="First Input", value=self.children[0].value, inline=False)
+        embed.add_field(name="Second Input", value=self.children[1].value, inline=False)
+        await interaction.response.send_message(embeds=[embed])
+
+
+@bot.slash_command(name="modaltest", guild_ids=[907533384081879040])
```
[the remark this answers is not in the document]
Comment (krittick) [reply examples/modal_dialogs.py]: Nothing to see there. :)

**Comment by krittick** [comment]
Since this feature is now released, let's prioritize reviewing this ASAP.

**Comment by Nziie3** [comment]
Wait, the Text Input Component has released? I thought it was just the form type things.

**Comment by krittick** [comment]
> Wait, the Text Input Component has released? I thought it was just the form type things. Or are the forms the text input components.

Text Input components are only available in modal dialogs.

**Review by Lulalaby — APPROVED** [review]

**Review by Middledot — APPROVED** [review]

**Pull request merged**